api/src/main/java/org/openmrs/util/OpenmrsUtil.java

```diff
@@ -624,10 +624,10 @@ public static int compare(Date d1, Date d2) {
 			return d1.compareTo(d2);
 		}
 		if (d1 instanceof Timestamp) {
-			d1 = new Date(((Timestamp) d1).getTime());
```

Comment:
Is Timestap.getTime() the same as Date.getTime()?

Reply:
Locally I have run a test
```java
import java.sql.Timestamp;
import java.util.Date;

@Test
void getTimeOfSqlTimestampVsJavaUtilDate() {
	long time = System.currentTimeMillis();
	Timestamp timestamp = new Timestamp(time);
	Date date = new Date(time);
	assertThat(timestamp.getTime()).isEqualTo(date.getTime());
}
```
and it's green

Reply:
Did you look at the actual implementations of the two methods?

Reply:
Yes, I know that the `Timestamp` has nanoseconds.
I have a feeling this code is doing something different then it is supposed to do. Please have a look.

There are two things happening here:
- not needed casts,
- logic that possibly does a different thing than intended

This PR is about a not needed cast:
```java
	if (d1 instanceof Timestamp) {
        // if we are here then it means that  d1 is an instance of a Timestamp
        // so why casting it into a Timestamp? Even without the cast, a Timestamp's getTime() will be called - not the getTime() from a parent class(Date)
        // it is polymorphism in action. 
		d1 = new Date(((Timestamp) d1).getTime());
	}
```
  
  
Now about the other point.
There are four scenarios based on the parameters passed to this method:
- `Date, Date`
- `Timestamp, Timestamp`,
- `Date, Timestamp`,
- `Timestamp, Date`

For `Date, Date` and `Timestamp, Timestamp` everything is ok.

Now let's check `d1=Date, d2=Timestamp`
```java
public static int compare(Date d1, Date d2) {
	if (d1 instanceof Timestamp && d2 instanceof Timestamp) {
        // not entered
	}

	if (d1 instanceof Timestamp) {
        // not entered
		d1 = new Date(((Timestamp) d1).getTime());
	}
	if (d2 instanceof Timestamp) {
        //d2.getTime() will return nanoseconds and they will be passed to a Date's constructor. This I think may create a date in the future..
        d2 = new Date(((Timestamp) d2).getTime());
	}

    //and because d2 is now a date in the future, this comparison will fail.
	return d1.compareTo(d2);
}
```
Reversed scenario, the one for `Timestamp, Date` will work the same.

But that's just my theory. Tests are needed to verify this so maybe let's do this:
- exclude these two casts from PR
- create a new task to provide tests covering those 4 scenarios and then we can come back to the idea of removing the casts from it. That's the safe way to go 💪 
What do you think about adding a separate task for writing tests for this method?

Reply:
I like your reasoning. Makes a lot of sense to me. 😊